File: externals/transportPAO/src/transport/hamiltonian.f90
!
! Copyright (C) 2005 WanT Group, 2017 ERMES Group
!
! This file is distributed under the terms of the
! GNU General Public License. See the file `License\'
! in the root directory of the present distribution,
! or http://www.gnu.org/copyleft/gpl.txt .
!
!*********************************************
   MODULE T_hamiltonian_module
   !*********************************************
   !
   USE kinds,                  ONLY : dbl
   USE constants,              ONLY : ZERO
   USE parameters,             ONLY : nstrx
   USE log_module,             ONLY : log_push, log_pop
   USE T_kpoints_module,       ONLY : nkpts_par
   USE T_operator_blc_module
   !
   IMPLICIT NONE
   PRIVATE 
   SAVE
!
! Contains the description of the system in terms of 
! Hamiltonian, Overlap, and correlation Sigma blocks
! 
    ! 
    ! dimensions
    !
    INTEGER                   :: dimL
    INTEGER                   :: dimC
    INTEGER                   :: dimR
    INTEGER                   :: dimx, dimx_lead
    !
    INTEGER                   :: nspin
    INTEGER                   :: ispin
    REAL(dbl)                 :: shift_L
    REAL(dbl)                 :: shift_C
    REAL(dbl)                 :: shift_R
    REAL(dbl)                 :: shift_corr
    !
    TYPE( operator_blc )      :: blc_00L 
    TYPE( operator_blc )      :: blc_01L 
    TYPE( operator_blc )      :: blc_00R 
    TYPE( operator_blc )      :: blc_01R 
    TYPE( operator_blc )      :: blc_00C
    TYPE( operator_blc )      :: blc_LC
    TYPE( operator_blc )      :: blc_CR
    !
    LOGICAL :: alloc = .FALSE.


!
! end delcarations
!

   PUBLIC :: dimL, dimC, dimR, dimx, dimx_lead
   PUBLIC :: nspin, ispin
   PUBLIC :: nkpts_par
   !
   PUBLIC :: shift_L, shift_C, shift_R, shift_corr
   !
   PUBLIC :: blc_00L
   PUBLIC :: blc_01L
   PUBLIC :: blc_00R
   PUBLIC :: blc_01R
   PUBLIC :: blc_00C
   PUBLIC :: blc_LC
   PUBLIC :: blc_CR
   !
   PUBLIC :: alloc
   !
   PUBLIC :: hamiltonian_allocate
   PUBLIC :: hamiltonian_deallocate
   PUBLIC :: hamiltonian_memusage


CONTAINS

!
! subroutines
!

!**********************************************************
   SUBROUTINE hamiltonian_allocate()
   !**********************************************************
   IMPLICIT NONE
      CHARACTER(20)      :: subname="hamiltonian_allocate"

      CALL log_push( subname )

      IF ( alloc )       CALL errore(subname,'already allocated', 1 )
      IF ( dimL <= 0 )   CALL errore(subname,'invalid dimL', 1 )
      IF ( dimR <= 0 )   CALL errore(subname,'invalid dimR', 1 )
      IF ( dimC <= 0 )   CALL errore(subname,'invalid dimC', 1 )
      IF ( nkpts_par <= 0 )   CALL errore(subname,'invalid nkpts_par', 1 )
      !
      dimx      = MAX( dimC, dimR, dimL )
      dimx_lead = MAX( dimR, dimL )

      !
      ! init data
      !
      CALL operator_blc_init( blc_00L, "block_00L")
      CALL operator_blc_init( blc_01L, "block_01L")
      CALL operator_blc_init( blc_00R, "block_00R")
      CALL operator_blc_init( blc_01R, "block_01R")
      CALL operator_blc_init( blc_00C, "block_00C")
      CALL operator_blc_init( blc_LC,  "block_LC")
      CALL operator_blc_init( blc_CR,  "block_CR")

      !
      ! allocations
      !
      CALL operator_blc_allocate( dimL, dimL, nkpts_par, OBJ=blc_00L )
      CALL operator_blc_allocate( dimL, dimL, nkpts_par, OBJ=blc_01L )
      !
      CALL operator_blc_allocate( dimR, dimR, nkpts_par, OBJ=blc_00R )
      CALL operator_blc_allocate( dimR, dimR, nkpts_par, OBJ=blc_01R )
      !
      CALL operator_blc_allocate( dimC, dimC, nkpts_par, OBJ=blc_00C )
      CALL operator_blc_allocate( dimL, dimC, nkpts_par, OBJ=blc_LC )
      CALL operator_blc_allocate( dimC, dimR, nkpts_par, OBJ=blc_CR )
      !
      alloc = .TRUE.

      CALL log_pop( subname )

   END SUBROUTINE hamiltonian_allocate


!**********************************************************
   SUBROUTINE hamiltonian_deallocate()
   !**********************************************************
   IMPLICIT NONE
      CHARACTER(22)      :: subname="hamiltonian_deallocate"

      CALL log_push( subname )

      IF ( .NOT. alloc ) RETURN

      IF ( blc_00L%alloc ) CALL operator_blc_deallocate( OBJ=blc_00L )
      IF ( blc_01L%alloc ) CALL operator_blc_deallocate( OBJ=blc_01L )
      !
      IF ( blc_00R%alloc ) CALL operator_blc_deallocate( OBJ=blc_00R )
      IF ( blc_01R%alloc ) CALL operator_blc_deallocate( OBJ=blc_01R )
      !
      IF ( blc_00C%alloc ) CALL operator_blc_deallocate( OBJ=blc_00C )
      IF ( blc_LC%alloc  ) CALL operator_blc_deallocate( OBJ=blc_LC )
      IF ( blc_CR%alloc  ) CALL operator_blc_deallocate( OBJ=blc_CR )
      !
      alloc = .FALSE.   

      CALL log_pop( subname )

   END SUBROUTINE hamiltonian_deallocate


!**********************************************************
   REAL(dbl) FUNCTION hamiltonian_memusage(memtype)
   !**********************************************************
   IMPLICIT NONE
       !
       CHARACTER(*)   :: memtype
       !
       REAL(dbl) :: cost
       !
       cost = ZERO
       IF ( blc_00L%alloc )    cost = cost + operator_blc_memusage(blc_00L,  memtype)
       IF ( blc_01L%alloc )    cost = cost + operator_blc_memusage(blc_01L,  memtype)
       IF ( blc_00R%alloc )    cost = cost + operator_blc_memusage(blc_00R,  memtype)
       IF ( blc_01R%alloc )    cost = cost + operator_blc_memusage(blc_01R,  memtype)
       IF ( blc_00C%alloc )    cost = cost + operator_blc_memusage(blc_00C,  memtype)
       IF ( blc_LC%alloc )     cost = cost + operator_blc_memusage(blc_LC,   memtype)
       IF ( blc_CR%alloc )     cost = cost + operator_blc_memusage(blc_CR,   memtype)
       !
       hamiltonian_memusage = cost
       !
   END FUNCTION hamiltonian_memusage


END MODULE T_hamiltonian_module

File: src/PAOFLOW/transport/hamiltonian/hamiltonian.py
from typing import Literal
import numpy as np
from PAOFLOW.transport.hamiltonian.operator_blc import OperatorBlock


class HamiltonianSystem:
    """
    Represents a transport system defined by block Hamiltonians, overlap matrices, and correlation self-energies.

    This class holds and manages all matrix blocks used in quantum transport setups including the left/right leads,
    the central conductor, and inter-region couplings. Each block is stored as an OperatorBlock instance.

    Methods are provided for memory allocation, deallocation, and estimating memory usage.

    Attributes
    ----------
    dimL : int
        Dimension of the left lead region.
    dimC : int
        Dimension of the central conductor region.
    dimR : int
        Dimension of the right lead region.
    nkpts_par : int
        Number of parallel k-points.
    shift_L : float
        Energy shift for the left region.
    shift_C : float
        Energy shift for the center region.
    shift_R : float
        Energy shift for the right region.
    shift_corr : float
        Energy shift for correlation self-energy.
    """

    def __init__(self, dimL: int, dimC: int, dimR: int, nkpts_par: int) -> None:
        self.dimL = dimL
        self.dimC = dimC
        self.dimR = dimR
        self.nkpts_par = nkpts_par

        self.shift_L = 0.0
        self.shift_C = 0.0
        self.shift_R = 0.0
        self.shift_corr = 0.0

        self.allocated = False
        self.dimx = max(dimL, dimC, dimR)
        self.dimx_lead = max(dimL, dimR)

        self.blc_00L = OperatorBlock("block_00L")
        self.blc_01L = OperatorBlock("block_01L")
        self.blc_00R = OperatorBlock("block_00R")
        self.blc_01R = OperatorBlock("block_01R")
        self.blc_00C = OperatorBlock("block_00C")
        self.blc_LC = OperatorBlock("block_LC")
        self.blc_CR = OperatorBlock("block_CR")

    def allocate(
        self,
        ivr_par: np.ndarray,
        tag_dict: dict[str, dict[str, str]],
    ) -> None:
        """
        Allocate memory for all matrix blocks and initialize their metadata.

        Parameters
        ----------
        ivr_par : np.ndarray
            2D array of shape (2, nrtot_par) specifying the parallel real-space R-vectors.
            This is used to match R-slices in the Hamiltonian during read-in.
        """
        if self.allocated:
            raise RuntimeError("Hamiltonian blocks already allocated.")
        if min(self.dimL, self.dimC, self.dimR, self.nkpts_par) <= 0:
            raise ValueError("Invalid dimensions for Hamiltonian allocation.")

        block_specs = [
            (self.blc_00L, self.dimL, self.dimL),
            (self.blc_01L, self.dimL, self.dimL),
            (self.blc_00R, self.dimR, self.dimR),
            (self.blc_01R, self.dimR, self.dimR),
            (self.blc_00C, self.dimC, self.dimC),
            (self.blc_LC, self.dimL, self.dimC),
            (self.blc_CR, self.dimC, self.dimR),
        ]

        for block, d1, d2 in block_specs:
            block.allocate(d1, d2, self.nkpts_par)
            block.tag = tag_dict.get(
                block.name,
                {
                    "rows": "all",
                    "cols": "all",
                    "rows_sgm": "all",
                    "cols_sgm": "all",
                },
            )
            block.ivr_par = ivr_par

        self.allocated = True

    def memusage(self, memtype: Literal["ham", "corr", "all"] = "all") -> float:
        """
        Estimate total memory usage of all blocks in MB.

        Parameters
        ----------
        memtype : {"ham", "corr", "all"}
            Type of memory to report.

        Returns
        -------
        usage_mb : float
            Estimated memory in megabytes.
        """
        usage = 0.0
        for block in self.blocks.values():
            if block.allocated:
                usage += block.memusage(memtype)
        return usage

    @property
    def blocks(self) -> dict[str, OperatorBlock]:
        return {
            "blc_00L": self.blc_00L,
            "blc_01L": self.blc_01L,
            "blc_00R": self.blc_00R,
            "blc_01R": self.blc_01R,
            "blc_00C": self.blc_00C,
            "blc_LC": self.blc_LC,
            "blc_CR": self.blc_CR,
        }
